src/components/molecule_3d.jsx

```diff
@@ -99,51 +136,30 @@ class Molecule3d extends React.Component {
   }
 
   render3dMol() {
-    const modelData = this.props.modelData;
-
-    if (!modelData.atoms.length || !modelData.bonds.length) {
+    if (!this.props.modelData.atoms.length ||
+        !this.props.modelData.bonds.length) {
       return;
     }
 
     const glviewer = this.glviewer || $3Dmol.createViewer(jQuery(this.container), {
       defaultcolors: $3Dmol.rasmolElementColors,
     });
 
-    glviewer.clear();
-
-    glviewer.addModel(moleculeUtils.modelDataToCDJSON(modelData), 'json', {
-      keepH: true,
-    });
-
-    // Hack in chain and residue data, since it's not supported by chemdoodle json
-    glviewer.getModel().selectedAtoms().forEach((atom) => {
-      const modifiedAtom = atom;
-      modifiedAtom.atom = modelData.atoms[atom.serial].name;
-      modifiedAtom.chain = modelData.atoms[atom.serial].chain;
-      modifiedAtom.resi = modelData.atoms[atom.serial].residue_index;
-      modifiedAtom.resn = modelData.atoms[atom.serial].residue_name;
-    });
-
-    const styles = this.props.styles;
-    modelData.atoms.forEach((atom, i) => {
-      const style = styles[i] || {};
-      const libStyle = {};
-      const visualizationType = style.visualization_type || DEFAULT_VISUALIZATION_TYPE;
-
-      libStyle[visualizationType] = {};
-      Object.keys(style).forEach((styleKey) => {
-        libStyle[visualizationType][styleKey] = style[styleKey];
-      });
-
-      if (this.state.selectedAtomIds.indexOf(atom.serial) !== -1) {
-        libStyle[visualizationType].color = 0x1FF3FE;
-      }
+    const renderingSameModelData = moleculeUtils.modelDataEquivalent(
+      this.oldModelData, this.props.modelData
+    );
+    if (!renderingSameModelData) {
+      this.lastStylesByAtom = null;
+      Molecule3d.render3dMolModel(glviewer, this.props.modelData);
+    }
 
-      if (typeof libStyle[visualizationType].color === 'string') {
-        libStyle[visualizationType].color = libUtils.colorStringToNumber(
-          libStyle[visualizationType].color
-        );
-      }
+    const styleUpdates = new Map(); // style update strings to atom ids needed
```

Comment:
This map allows us to prevent setting a style to an atom that already has that style.  It keeps track of the current style that has been set to each atom, so if a new style matches we don't set it again.